Add ILIKE functionality

alireza-ahmadi opened the pull request:
### Description

Loopback-datasource-juggler added support for `ILIKE` and `NOT ILIKE` functionality in **v3.1.0**, but the PostgreSQL connector does not update to support them. This pull request adds this functionality by adding a few lines to `lib/postgresql.js` file.

#### Related issues

- strongloop/loopback-datasource-juggler#1091

Please let me if any other changes needed.

slnode commented:
Can one of the admins verify this patch? To accept patch and trigger a build add comment ".*ok\W+to\W+test.*"

slnode commented:
Can one of the admins verify this patch?

slnode commented:
Can one of the admins verify this patch?

slnode commented:
Can one of the admins verify this patch?

superkhau commented:
Can you add some tests to verify your changes and prevent regressions in the future?

alireza-ahmadi commented:
Yes, I will add a few test in the next few days.

zbarbuto commented:
@alireza-ahmadi any news on this?

alireza-ahmadi commented:
@zbarbuto @superkhau I have added a few tests for the added feature. Also because I couldn't find the required test for `LIKE` and `NOT LIKE` operators, I have added a few tests for them as well because I thought that all of these operators are pattern matching operators and are linked.

Please let me know if the tests don't have a good coverage or any other changes needed.

superkhau commented on test/postgresql.test.js:
```diff
@@ -230,6 +230,121 @@ describe('postgresql connector', function() {
+      function(done) {
+        Post.find({where: {content: {like: '%TestCase%'}}},
+          function(err, posts) {
+            should.not.exists(err);
```
`if (err) return done(err);` for better stack traces

superkhau commented on test/postgresql.test.js:
```diff
@@ -230,6 +230,121 @@ describe('postgresql connector', function() {
+      Post.destroyAll(done);
+    });
+
+    it('should support case sensitive queries using like',
```
it('supports case sensi...

superkhau replied:
Yes, I see the other tests are different (we need to eventually update all of them to this style over time

superkhau commented on test/postgresql.test.js:
```diff
@@ -230,6 +230,121 @@ describe('postgresql connector', function() {
+          });
+      });
+
+    it('should not support case insensitive queries using like',
```
ditto

superkhau commented on test/postgresql.test.js:
```diff
@@ -230,6 +230,121 @@ describe('postgresql connector', function() {
+      function(done) {
+        Post.find({where: {content: {like: '%tesTcasE%'}}},
+          function(err, posts) {
+            should.not.exists(err);
```
ditto

superkhau commented on test/postgresql.test.js:
```diff
@@ -230,6 +230,121 @@ describe('postgresql connector', function() {
+          });
+      });
+
+    it('should support like for no match', function(done) {
```
ditto

superkhau commented on test/postgresql.test.js:
```diff
@@ -230,6 +230,121 @@ describe('postgresql connector', function() {
+    it('should support like for no match', function(done) {
+      Post.find({where: {content: {like: '%TestXase%'}}},
+        function(err, posts) {
+          should.not.exists(err);
```
ditto

superkhau commented on test/postgresql.test.js:
```diff
@@ -230,6 +230,121 @@ describe('postgresql connector', function() {
+        });
+    });
+
+    it('should support negative case sensitive queries using nlike',
```
ditto

superkhau commented on test/postgresql.test.js:
```diff
@@ -230,6 +230,121 @@ describe('postgresql connector', function() {
+      function(done) {
+        Post.find({where: {content: {nlike: '%Case%'}}},
+          function(err, posts) {
+            should.not.exists(err);
```
ditto

superkhau commented on test/postgresql.test.js:
```diff
@@ -230,6 +230,121 @@ describe('postgresql connector', function() {
+          });
+      });
+
+    it('should not support negative case insensitive queries using nlike',
```
ditto

superkhau commented on test/postgresql.test.js:
```diff
@@ -230,6 +230,121 @@ describe('postgresql connector', function() {
+      function(done) {
+        Post.find({where: {content: {nlike: '%casE%'}}},
+          function(err, posts) {
+            should.not.exists(err);
```
ditto

superkhau commented on test/postgresql.test.js:
```diff
@@ -230,6 +230,121 @@ describe('postgresql connector', function() {
+          });
+      });
+
+    it('should support nlike for no match', function(done) {
```
ditto

superkhau commented on test/postgresql.test.js:
```diff
@@ -230,6 +230,121 @@ describe('postgresql connector', function() {
+    it('should support nlike for no match', function(done) {
+      Post.find({where: {content: {nlike: '%TestXase%'}}},
+        function(err, posts) {
+          should.not.exists(err);
```
ditto

superkhau commented on test/postgresql.test.js:
```diff
@@ -230,6 +230,121 @@ describe('postgresql connector', function() {
+        });
+    });
+
+    it('should support case insensitive queries using ilike', function(done) {
```
ditto

superkhau commented on test/postgresql.test.js:
```diff
@@ -230,6 +230,121 @@ describe('postgresql connector', function() {
+    it('should support case insensitive queries using ilike', function(done) {
+      Post.find({where: {content: {ilike: '%tesTcasE%'}}},
+        function(err, posts) {
+          should.not.exist(err);
```
ditto

superkhau commented on test/postgresql.test.js:
```diff
@@ -230,6 +230,121 @@ describe('postgresql connector', function() {
+        });
+    });
+
+    it('should support ilike for no match', function(done) {
```
ditto

superkhau commented on test/postgresql.test.js:
```diff
@@ -230,6 +230,121 @@ describe('postgresql connector', function() {
+    it('should support ilike for no match', function(done) {
+      Post.find({where: {content: {ilike: '%tesTxasE%'}}},
+        function(err, posts) {
+          should.not.exists(err);
```
ditto

superkhau commented on test/postgresql.test.js:
```diff
@@ -230,6 +230,121 @@ describe('postgresql connector', function() {
+        });
+    });
+
+    it('should support negative case insensitive queries using nilike',
```
ditto

superkhau commented on test/postgresql.test.js:
```diff
@@ -230,6 +230,121 @@ describe('postgresql connector', function() {
+      function(done) {
+        Post.find({where: {content: {nilike: '%casE%'}}},
+          function(err, posts) {
+            should.not.exist(err);
```
ditto

superkhau commented on test/postgresql.test.js:
```diff
@@ -230,6 +230,121 @@ describe('postgresql connector', function() {
+          });
+      });
+
+    it('should support nilike for no match', function(done) {
```
ditto

superkhau commented on test/postgresql.test.js:
```diff
@@ -230,6 +230,121 @@ describe('postgresql connector', function() {
+    it('should support nilike for no match', function(done) {
+      Post.find({where: {content: {nilike: '%tesTxasE%'}}},
+        function(err, posts) {
+          should.not.exists(err);
```
ditto

alireza-ahmadi commented:
@superkhau Thanks for the review, I have applied the requested changes. Please let me know if any other changes required.

superkhau commented:
@slnode test please

superkhau commented on test/postgresql.test.js:
```diff
@@ -230,6 +230,111 @@ describe('postgresql connector', function() {
+    });
+    after(function deleteTestFixtures(done) {
+      Post.destroyAll(done);
+    });
```
You can remove the after since we delete in the before already

alireza-ahmadi replied:
@superkhau Thanks for the review. I will apply these changes, rebase, and push the changes in the next few minutes.

superkhau commented on test/postgresql.test.js:
```diff
@@ -230,6 +230,111 @@ describe('postgresql connector', function() {
+        title: 't2',
+        content: 'T2_TheOtherCase',
+      }], done);
+    });
```
You can move the two functions to the bottom of the context block:

```
before(deleteTestFixtures);
before(createTestFixtures);

it('...');
it('...');
...

function deleteTestFixtures(done) {
  Post.destroyAll(done);
}
function create...
```

http://loopback.io//doc/en/contrib/style-guide.html#layout-of-test-files

superkhau commented:
@kjdelisle Reassigning to your squad.

alireza-ahmadi commented:
I have applied the requested code style changes(and a typo in `createTestFixtures`) and rebased to the last state of `master` branch. Again, please let me know if any other changes required.

superkhau commented:
Let's wait for CI to pass, should be good to land.

zbarbuto commented:
Thanks for your effort @alireza-ahmadi 

@superkhau Can we expect a new point release soon with this feature?

superkhau commented:
@zbarbuto Can be released as soon as it lands. @kjdelisle will help you there once it passes CI -- I think it's hanging ATM now though. ;(

kjdelisle commented:
@slnode test please

kjdelisle commented:
LGTM
